In [352]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn import linear_model
from sklearn import model_selection
from sklearn import metrics
from sklearn import preprocessing
from sklearn import feature_selection
from sklearn import tree
from sklearn import ensemble

pd.set_option('display.float_format', lambda x: '%.5f' % x)

In [353]:
df0=pd.read_csv("F:\Programming\Machine Learning\ClassWorks\project\interview.csv")

In [354]:
#pd.set_option("display.max_rows", None, "display.max_columns", None)

In [355]:
df0.head()

Date of Interview Client name         Industry Location  \
0        13.02.2015     Hospira  Pharmaceuticals  Chennai   
1        13.02.2015     Hospira  Pharmaceuticals  Chennai   
2        13.02.2015     Hospira  Pharmaceuticals  Chennai   
3        13.02.2015     Hospira  Pharmaceuticals  Chennai   
4        13.02.2015     Hospira  Pharmaceuticals  Chennai   

  Position to be closed Nature of Skillset    Interview Type Name(Cand ID)  \
0   Production- Sterile            Routine  Scheduled Walkin   Candidate 1   
1   Production- Sterile            Routine  Scheduled Walkin   Candidate 2   
2   Production- Sterile            Routine  Scheduled Walkin   Candidate 3   
3   Production- Sterile            Routine  Scheduled Walkin   Candidate 4   
4   Production- Sterile            Routine  Scheduled Walkin   Candidate 5   

  Gender Candidate Current Location  ...  \
0   Male                    Chennai  ...   
1   Male                    Chennai  ...   
2   Male                    Chennai  ...   
3   Male                    Chennai  ...   
4   Male                    Chennai  ...   

  Are you clear with the venue details and the landmark.  \
0                                                Yes       
1                                                Yes       
2                                                NaN       
3                                                Yes       
4                                                Yes       

  Has the call letter been shared Expected Attendance Observed Attendance  \
0                             Yes                 Yes                  No   
1                             Yes                 Yes                  No   
2                             NaN           Uncertain                  No   
3                             Yes           Uncertain                  No   
4                             Yes           Uncertain                  No   

  Marital Status Unnamed: 23 Unnamed: 24 Unnamed: 25 Unnamed: 26 Unnamed: 27  
0         Single         NaN         NaN         NaN         NaN         NaN  
1         Single         NaN         NaN         NaN         NaN         NaN  
2         Single         NaN         NaN         NaN         NaN         NaN  
3         Single         NaN         NaN         NaN         NaN         NaN  
4        Married         NaN         NaN         NaN         NaN         NaN  

[5 rows x 28 columns]

In [356]:
df0=df0.drop(["Unnamed: 23","Unnamed: 24","Unnamed: 25", "Unnamed: 26", "Unnamed: 27"],axis=1)
df0=df0.drop(1233)
df0=df0.drop(["Date of Interview","Name(Cand ID)"],axis=1)

In [357]:
df0.shape

(1233, 21)

In [358]:
#Renaming columns
df1 = df0.rename(columns = {'Date of Interview':'Interview_date', 'Client name':'Client_name', 'Industry':'Industry', 'Location':'Location','Position to be closed':'Closed_pos' , 'Nature of Skillset':'Skillset', 'Interview Type':'Interview_type','Name(Cand ID)':'Cand_id', 'Gender':'Gender', 'Candidate Current Location':'Curr_loc','Candidate Job Location':'Curr_job_loc', 'Interview Venue':'Interview_venue','Candidate Native location':'Native_loc','Have you obtained the necessary permission to start at the required time':'Permission_status','Hope there will be no unscheduled meetings':'Unscd_meet_status','Can I Call you three hours before the interview and follow up on your attendance for the interview':'Followup_status','Can I have an alternative number/ desk number. I assure you that I will not trouble you too much':'AltNum_status','Have you taken a printout of your updated resume. Have you read the JD and understood the same':'Resume_status','Are you clear with the venue details and the landmark.':'Venue_status','Has the call letter been shared':'Call_letter_status', 'Expected Attendance':'Ex_attend','Observed Attendance':'Obs_attend', 'Marital Status':'Matital_status'}, inplace = False)

In [359]:
df1.columns

Index(['Client_name', 'Industry', 'Location', 'Closed_pos', 'Skillset',
       'Interview_type', 'Gender', 'Curr_loc', 'Curr_job_loc',
       'Interview_venue', 'Native_loc', 'Permission_status',
       'Unscd_meet_status', 'Followup_status', 'AltNum_status',
       'Resume_status', 'Venue_status', 'Call_letter_status', 'Ex_attend',
       'Obs_attend', 'Matital_status'],
      dtype='object')

In [360]:
df1.shape

(1233, 21)

# Cleaning the DATA

### Column 1. "Client_name"

In [361]:
df1["Client_name"].value_counts(),df1["Client_name"].value_counts().sum()

(Standard Chartered Bank            904
 Pfizer                              75
 Hospira                             75
 Aon Hewitt                          28
 Flextronics                         23
 ANZ                                 22
 Hewitt                              20
 UST                                 18
 Standard Chartered Bank Chennai     17
 Prodapt                             17
 Astrazeneca                         15
 Williams Lea                        11
 Barclays                             5
 Aon hewitt Gurgaon                   2
 Woori Bank                           1
 Name: Client_name, dtype: int64,
 1233)

In [362]:
df1["Client_name"].isna().sum()#NO missing values

0

### Column 2. "Industry"

In [363]:
df1["Industry"].value_counts(),df1["Industry"].value_counts().sum()

(BFSI                        949
 Pharmaceuticals             165
 IT Products and Services     45
 Electronics                  23
 IT Services                  23
 Telecom                      17
 IT                           11
 Name: Industry, dtype: int64,
 1233)

In [364]:
df1["Industry"].isna().sum()#NO missing values

0

In [365]:
# Group IT in one category
def rename_industry(data):
    if 'IT' in data:
        return 'IT'
    else:
        return data
df1.Industry=df1.Industry.apply(rename_industry)

In [366]:
df1["Industry"].value_counts(),df1["Industry"].value_counts().sum()

(BFSI               949
 Pharmaceuticals    165
 IT                  79
 Electronics         23
 Telecom             17
 Name: Industry, dtype: int64,
 1233)

### Column 3. "Location"

In [367]:
df1["Location"].value_counts(),df1["Location"].value_counts().sum()

(Chennai       754
 Bangalore     292
 chennai        86
 Hyderabad      38
 Gurgaon        33
 Noida          15
 - Cochin-       9
 chennai         3
 Delhi           1
 Gurgaonr        1
 CHENNAI         1
 Name: Location, dtype: int64,
 1233)

In [368]:
#NOTICE that, the value of this column are similar to that of column "Curr_loc".
#Therefore we can simply drop this column.
df1=df1.drop("Location",axis=1)

### Column 4. "Closed_pos"

In [369]:
df1["Closed_pos"].value_counts(),df1["Closed_pos"].value_counts().sum()

(Routine                1023
 Niche                   163
 Dot Net                  18
 Trade Finance            11
 AML                       8
 Selenium testing          5
 Production- Sterile       5
 Name: Closed_pos, dtype: int64,
 1233)

In [370]:
df1["Closed_pos"].isna().sum()#NO missing values

0

In [371]:
#Since Routine values are significantly more compared to other values.
#Grouping Closed position into Routine or Non-Routine.
def rename_Closed_pos(data):
    value=data.rstrip().lower()
    if value == 'routine':
        return 'Routine'
    else:
        return 'Non-Routine'

df1.Closed_pos=df1.Closed_pos.apply(rename_Closed_pos)

In [372]:
df1["Closed_pos"].value_counts(),df1["Closed_pos"].value_counts().sum()

(Routine        1023
 Non-Routine     210
 Name: Closed_pos, dtype: int64,
 1233)

### Column 5. "Skillset"

In [373]:
df1["Skillset"].value_counts(),df1["Skillset"].value_counts().sum()

(JAVA/J2EE/Struts/Hibernate    220
 Accounting Operations          86
 Fresher                        86
 AML/KYC/CDD                    84
 CDD KYC                        52
                              ... 
 SCCM – Sharepoint               1
 sccm                            1
 11.30 Am                        1
 JAVA, J2ee                      1
 BaseSAS Program/ Reporting      1
 Name: Skillset, Length: 92, dtype: int64,
 1233)

In [374]:
df1=df1[~(df1["Skillset"]== '11.30 AM')]

df1=df1[~(df1["Skillset"]== '10.00 AM')]

df1=df1[~(df1["Skillset"]== '11.30 Am')]

df1=df1[~(df1["Skillset"]== '9.30 AM')]

df1=df1[~(df1["Skillset"]== '9.00 Am')]

df1=df1[~(df1["Skillset"]== '12.30 Pm')]

In [375]:
df1.shape

(1226, 20)

In [376]:
df1["Skillset"].value_counts(),df1["Skillset"].value_counts().sum()

(JAVA/J2EE/Struts/Hibernate    220
 Accounting Operations          86
 Fresher                        86
 AML/KYC/CDD                    84
 CDD KYC                        52
                              ... 
 Biosimilars                     1
 SCCM – Sharepoint               1
 sccm                            1
 JAVA, J2ee                      1
 BaseSAS Program/ Reporting      1
 Name: Skillset, Length: 86, dtype: int64,
 1226)

In [377]:
df1["Skillset"].nunique()

86

In [378]:
df1["Skillset"].isna().sum()#NO missing values

0

In [379]:
df1["Skillset"].value_counts().sum()

1226

In [380]:
#Merging 'Sillset' data and creating new columns "Skill" which is cleaner.
df1['Skill']=pd.Series()
df1['Skill'][df1['Skillset'].str.contains("eveloper")]='Developer'
df1['Skill'][df1['Skillset'].str.contains("SAP")]='SAP'
df1['Skill'][df1['Skillset'].str.contains("naly")]='Analyst'
df1['Skill'][df1['Skillset'].str.contains("RA")]='RA'
df1['Skill'][df1['Skillset'].str.contains("roduc")]='Product'
df1['Skill'][df1['Skillset'].str.contains("SCC")]='SCCM'
df1['Skill'][df1['Skillset'].str.contains("Scc")]='SCCM'
df1['Skill'][df1['Skillset'].str.contains("sccm")]='SCCM'
df1['Skill'][df1['Skillset'].str.contains("abili")]='LL'
df1['Skill'][df1['Skillset'].str.contains("L & L")]='LL'
df1['Skill'][df1['Skillset'].str.contains("abli")]='LL'
df1['Skill'][df1['Skillset'].str.contains("Publi")]='Publishing'
df1['Skill'][df1['Skillset'].str.contains('esti')]='Testing'
df1['Skill'][df1['Skillset'].str.contains("ednet")]='Mednet'
df1['Skill'][df1['Skillset'].str.contains("Java")]='Java'
df1['Skill'][df1['Skillset'].str.contains("JAVA")]='Java'
df1['Skill'][df1['Skillset'].str.contains("Oracle")]='Oracle'
df1['Skill'][df1['Skillset'].str.contains("Fres")]='Fresher'
df1['Skill'][df1['Skillset'].str.contains("pera")]='Operations'
df1['Skill'][df1['Skillset'].str.contains("KYC")]='KYC'
df1['Skill'][df1['Skillset'].str.contains("Routi")]='Routine'
df1['Skill'][df1['Skillset'].str.contains("SAS")]='SAS'
df1['Skill'][df1['Skillset'].str.contains("anag")]='Manager'
df1['Skill'][df1['Skillset'].str.contains("iosim")]='Biosimilarity'
df1['Skill'][df1['Skillset'].str.contains("ot ")]='Dot Net'
df1['Skill'][df1['Skillset'].str.contains("Hadoop")]='Hadoop'
df1['Skill'][df1['Skillset'].str.contains("egula")]='Regulatory'
df1['Skill'][df1['Skillset'].str.contains("EMEA")]='EMEA'
df1['Skill'][df1['Skillset'].str.contains("COTS")]='COTS'
df1['Skill'][df1['Skillset'].str.contains("ETL")]='ETL'
df1['Skill'][df1['Skillset'].str.contains("abel")]='Labelling'

<ipython-input-380-807a17c05c07>:2: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  df1['Skill']=pd.Series()
<ipython-input-380-807a17c05c07>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Skill'][df1['Skillset'].str.contains("eveloper")]='Developer'


In [381]:
df1["Skill"].value_counts().sum()

1222

In [382]:
df1["Skill"].value_counts()

Java             454
KYC              136
Operations       110
Fresher           86
Oracle            68
Routine           47
Testing           39
SAS               33
LL                33
SCCM              25
Mednet            24
Analyst           21
Publishing        18
COTS              17
Developer         15
Product           13
Regulatory        12
Hadoop            12
Biosimilarity     10
Dot Net            9
ETL                9
RA                 8
Labelling          8
EMEA               6
Manager            5
SAP                4
Name: Skill, dtype: int64

In [383]:
#droping the 'Skillset' column
df1=df1.drop("Skillset",axis=1)

In [384]:
df1["Skill"].isna().sum()

4

In [385]:
#set missing values as 'others'
df1["Skill"]=df1["Skill"].fillna(value='others')

In [386]:
df1.shape

(1226, 20)

In [387]:
df1["Skill"].isna().sum()

0

In [388]:
df1["Skill"].nunique()

27

In [389]:
#droping rows which has less distinct values 
df1.drop(df1.loc[df1['Skill']=='SAP'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Manager'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='EMEA'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Labelling'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='RA'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='ETL'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Dot Net'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Biosimilarity'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Hadoop'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Regulatory'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Product'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Developer'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='COTS'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Publishing'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Analyst'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='Mednet'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='SCCM'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='LL'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='SAS'].index, inplace=True)
df1.drop(df1.loc[df1['Skill']=='others'].index, inplace=True)

In [390]:
df1['Skill'].value_counts()

Java          454
KYC           136
Operations    110
Fresher        86
Oracle         68
Routine        47
Testing        39
Name: Skill, dtype: int64

### Column 6. "Interview_type"

In [391]:
df1["Interview_type"].value_counts(),df1["Interview_type"].value_counts().sum()

(Scheduled Walk In    449
 Scheduled Walkin     168
 Walkin               167
 Scheduled            128
 Walkin                27
 Sceduled walkin        1
 Name: Interview_type, dtype: int64,
 940)

In [392]:
df1["Interview_type"].isna().sum()#NO missing values

0

In [393]:
#re-classifying them into walkin, scheduled and scheduled_walkin
def rename_interview_type(data):
    interview_type=data.rstrip().lower()
    if interview_type=='walkin':
        return 'walkin'
    elif interview_type =='scheduled':
        return 'scheduled'
    else:
        return 'scheduled_walkin'
df1.Interview_type=df1.Interview_type.apply(rename_interview_type) 
    

In [394]:
df1["Interview_type"].value_counts()

scheduled_walkin    618
walkin              194
scheduled           128
Name: Interview_type, dtype: int64

In [395]:
df1["Interview_type"].value_counts(),df1["Interview_type"].value_counts().sum()

(scheduled_walkin    618
 walkin              194
 scheduled           128
 Name: Interview_type, dtype: int64,
 940)

### Column 7. "Gender"

In [396]:
df1["Gender"].value_counts(),df1["Gender"].value_counts().sum()

(Male      723
 Female    217
 Name: Gender, dtype: int64,
 940)

In [397]:
df1["Gender"].isna().sum()#NO missing values

0

### Column 8."Curr_loc"

In [398]:
df1["Curr_loc"].value_counts(),df1["Curr_loc"].value_counts().sum()

(Chennai      605
 Bangalore    227
 chennai       76
 Gurgaon       20
 Noida          8
 chennai        3
 Delhi          1
 Name: Curr_loc, dtype: int64,
 940)

In [399]:
df1["Curr_loc"].isna().sum()#NO missing values

0

In [400]:
#Merging the duplicate datas
def rename_cochin(data):
    value=data.rstrip().lower()
    if value=='- cochin-':
        return 'Cochin'
    elif value=='chennai':
        return 'Chennai'
    else:
        return value
df1.Curr_loc=df1.Curr_loc.apply(rename_cochin) 

In [401]:
df1["Curr_loc"].value_counts()

Chennai      684
bangalore    227
gurgaon       20
noida          8
delhi          1
Name: Curr_loc, dtype: int64

### Column 9."Curr_job_loc"

In [402]:
df1["Curr_job_loc"].value_counts(),df1["Curr_job_loc"].value_counts().sum()

(Chennai      681
 Bangalore    229
 Gurgaon       21
 Noida          8
 Hosur          1
 Name: Curr_job_loc, dtype: int64,
 940)

In [403]:
df1["Curr_job_loc"].isna().sum()#NO missing values

0

In [404]:
#Removing ' - ' from '- Cochin- ' values
def rename_Curr_job_loc(data):
    value=data.rstrip().lower()
    if value=='- cochin-':
        return 'Cochin'
    else:
        return value
df1.Curr_job_loc=df1.Curr_job_loc.apply(rename_Curr_job_loc) 

In [405]:
df1["Curr_job_loc"].value_counts(),df1["Curr_job_loc"].value_counts().sum()

(chennai      681
 bangalore    229
 gurgaon       21
 noida          8
 hosur          1
 Name: Curr_job_loc, dtype: int64,
 940)

### Column 10."Interview_venue"

In [406]:
df1["Interview_venue"].value_counts(),df1["Interview_venue"].value_counts().sum()

(Chennai      679
 Bangalore    227
 Gurgaon       21
 Noida          8
 Hosur          5
 Name: Interview_venue, dtype: int64,
 940)

In [407]:
df1["Interview_venue"].isna().sum()#NO missing values

0

In [408]:
#Removing ' - ' from '- Cochin- ' values
def rename_Interview_venue(data):
    value=data.rstrip().lower()
    if value=='- cochin-':
        return 'Cochin'
    else:
        return value
df1.Interview_venue=df1.Interview_venue.apply(rename_Interview_venue) 

In [409]:
df1["Interview_venue"].value_counts(),df1["Interview_venue"].value_counts().sum()

(chennai      679
 bangalore    227
 gurgaon       21
 noida          8
 hosur          5
 Name: Interview_venue, dtype: int64,
 940)

### Column 11."Native_loc"

In [410]:
df1["Native_loc"].value_counts(),df1["Native_loc"].value_counts().sum()

(Chennai          470
 Bangalore        131
 Hyderabad        124
 Cochin            23
 Allahabad         18
 Pune              17
 Coimbatore        17
 Cuttack           16
 Gurgaon           15
 Nagercoil         12
 Trivandrum        10
 Vellore           10
 Trichy            10
 Kolkata            9
 Visakapatinam      8
 Noida              8
 Mumbai             5
 Chandigarh         5
 Chitoor            4
 Pondicherry        4
 Hosur              3
 Tirupati           3
 Ahmedabad          2
 Delhi /NCR         2
 Delhi              2
 Patna              2
 Tuticorin          1
 Panjim             1
 Agra               1
 Tanjore            1
 Lucknow            1
 Salem              1
 Belgaum            1
 Mysore             1
 Kanpur             1
 Faizabad           1
 Name: Native_loc, dtype: int64,
 940)

In [411]:
df1["Native_loc"].nunique()

36

In [412]:
#we can drop the column 'Native_loc' as we dont think it is not that significant and also many distict values.
df1=df1.drop("Native_loc",axis=1)

### Column 12."Permission_status"

In [413]:
df1["Permission_status"].value_counts()

Yes               716
No                 53
Not yet            11
Yet to confirm      4
yes                 3
NO                  1
Name: Permission_status, dtype: int64

In [414]:
#Merging "Permission_status" values
def rename_permission(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'not yet' or value == 'na' or value =='no':
        return 'no'
    elif value == 'yet to confirm' or value == 'yes':
        return 'yes'
    else:
        return data

df1.Permission_status = df1.Permission_status.apply(rename_permission)

In [415]:
df1["Permission_status"].isna().sum()

152

In [416]:
#Consider missing values as 'no'
df1["Permission_status"]=df1["Permission_status"].fillna(value='no')

In [417]:
df1["Permission_status"].value_counts(),df1["Permission_status"].value_counts().sum()

(yes    723
 no     217
 Name: Permission_status, dtype: int64,
 940)

In [418]:
df1["Permission_status"].isna().sum()

0

In [419]:
df1.head(4)

Client_name         Industry   Closed_pos    Interview_type Gender Curr_loc  \
0     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
1     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
2     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
3     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   

  Curr_job_loc Interview_venue Permission_status Unscd_meet_status  \
0        hosur           hosur               yes               Yes   
1    bangalore           hosur               yes               Yes   
2      chennai           hosur                no                Na   
3      chennai           hosur               yes               Yes   

  Followup_status AltNum_status Resume_status Venue_status Call_letter_status  \
0             Yes           Yes           Yes          Yes                Yes   
1             Yes           Yes           Yes          Yes                Yes   
2             NaN           NaN           NaN          NaN                NaN   
3              No           Yes            No          Yes                Yes   

   Ex_attend Obs_attend Matital_status    Skill  
0        Yes         No         Single  Routine  
1        Yes         No         Single  Routine  
2  Uncertain         No         Single  Routine  
3  Uncertain         No         Single  Routine

### Column 13."Unscd_meet_status"

In [420]:
df1["Unscd_meet_status"].value_counts(),df1["Unscd_meet_status"].value_counts().sum()

(Yes         737
 Na           11
 yes           4
 No            2
 cant Say      1
 Not sure      1
 Name: Unscd_meet_status, dtype: int64,
 756)

In [421]:
#Merging "Unscd_meet_status" values
def rename_Unscd_meet_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'unsure' or value == 'not sure' or value == 'cant say' or value == 'nan' or value == 'na':
        return 'no'
    else:
        return 'yes'

df1.Unscd_meet_status=df1.Unscd_meet_status.apply(rename_Unscd_meet_status)

In [422]:
df1["Unscd_meet_status"].value_counts(),df1["Unscd_meet_status"].value_counts().sum()

(yes    743
 no      13
 Name: Unscd_meet_status, dtype: int64,
 756)

In [423]:
df1["Unscd_meet_status"].isna().sum()

184

In [424]:
#Consider missing values as 'no'
df1["Unscd_meet_status"] = df1["Unscd_meet_status"].fillna(value='no')

In [425]:
df1["Unscd_meet_status"].value_counts(),df1["Unscd_meet_status"].value_counts().sum()

(yes    743
 no     197
 Name: Unscd_meet_status, dtype: int64,
 940)

In [426]:
df1["Unscd_meet_status"].isna().sum()

0

### Column 14."AltNum_status"

In [427]:
df1["AltNum_status"].value_counts(),df1["AltNum_status"].value_counts().sum()

(Yes                          723
 No                            20
 Na                            10
 No I have only thi number      2
 na                             1
 Name: AltNum_status, dtype: int64,
 756)

In [428]:
df1["AltNum_status"].isna().sum()

184

In [429]:
#Merging "AltNum_status" values
def rename_AltNum_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'no i have only this number' or value == 'na':
        return 'no'
    else:
        return 'yes'

df1.AltNum_status=df1.AltNum_status.apply(rename_AltNum_status)

In [430]:
#Consider missing values as 'no'
df1["AltNum_status"]=df1["AltNum_status"].fillna(value='no')

In [431]:
df1["AltNum_status"].value_counts(),df1["AltNum_status"].value_counts().sum()

(yes    745
 no     195
 Name: AltNum_status, dtype: int64,
 940)

In [432]:
df1["AltNum_status"].isna().sum()

0

### Column 15."Resume_status"

In [433]:
df1["Resume_status"].value_counts(),df1["Resume_status"].value_counts().sum()

(Yes                      730
 No                        10
 Na                        10
 Not yet                    2
 na                         1
 No- will take it soon      1
 Not Yet                    1
 Name: Resume_status, dtype: int64,
 755)

In [434]:
df1["Resume_status"].isna().sum()

185

In [435]:
#Merging "Resume_status" values
def rename_Resume_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'no- will take it soon' or value == 'not yet' or value == 'na' or value=='no':
        return 'no'
    elif value == 'yes':
        return 'yes'
    
df1.Resume_status=df1.Resume_status.apply(rename_Resume_status)

In [436]:
#Consider missing values as 'no'
df1["Resume_status"]=df1["Resume_status"].fillna(value='no')

In [437]:
df1["Resume_status"].value_counts(),df1["Resume_status"].value_counts().sum()

(yes    730
 no     210
 Name: Resume_status, dtype: int64,
 940)

In [438]:
df1["Resume_status"].isna().sum()

0

### Column 16."Venue_status"

In [439]:
df1["Venue_status"].value_counts(),df1["Venue_status"].value_counts().sum()

(Yes                    735
 Na                      10
 No                       7
 No- I need to check      2
 na                       1
 Name: Venue_status, dtype: int64,
 755)

In [440]:
df1["Venue_status"].isna().sum()

185

In [441]:
#Merging "Venue_status" values
def rename_Venue_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'no- i need to check' or value == 'not yet' or value == 'na' or value=='no':
        return 'no'
    elif value == 'yes':
        return 'yes'
    
df1.Venue_status=df1.Venue_status.apply(rename_Venue_status)

In [442]:
#Consider missing values as 'no'
df1["Venue_status"] = df1["Venue_status"].fillna(value='no')

In [443]:
df1["Venue_status"].value_counts(),df1["Venue_status"].value_counts().sum()

(yes    735
 no     205
 Name: Venue_status, dtype: int64,
 940)

In [444]:
df1["Venue_status"].isna().sum()

0

### Column 17."Followup_status"

In [445]:
df1["Followup_status"].value_counts(),df1["Followup_status"].value_counts().sum()

(Yes    736
 Na      11
 No       7
 yes      2
 Name: Followup_status, dtype: int64,
 756)

In [446]:
df1["Followup_status"].isna().sum()

184

In [447]:
#Merging "Followup_status" values
def rename_Followup_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'no dont' or value == 'no' or value == 'na':
        return 'no'
    elif value == 'yes':
        return 'yes'
    
df1.Followup_status=df1.Followup_status.apply(rename_Followup_status)

In [448]:
#Consider missing values as 'no'
df1["Followup_status"] = df1["Followup_status"].fillna(value='no')

In [449]:
df1["Followup_status"].value_counts(),df1["Followup_status"].value_counts().sum()

(yes    738
 no     202
 Name: Followup_status, dtype: int64,
 940)

In [450]:
df1["Followup_status"].isna().sum()

0

### Column 18."Call_letter_status"

In [451]:
df1["Call_letter_status"].value_counts(),df1["Call_letter_status"].value_counts().sum()

(Yes               723
 No                 10
 Na                 10
 Not Sure            4
 Need To Check       3
 na                  1
 no                  1
 Not yet             1
 Yet to Check        1
 Havent Checked      1
 Not sure            1
 Name: Call_letter_status, dtype: int64,
 756)

In [452]:
df1["Call_letter_status"].isna().sum()

184

In [453]:
#Merging "Call_letter_status" values
def rename_Call_letter_status(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value == 'not sure' or value == 'no' or value == 'need to check' or value == "not yet" or value == "yet to check" or value == "havent checked" or value == "na":
        return 'no'
    elif value == 'yes':
        return 'yes'
    
df1.Call_letter_status=df1.Call_letter_status.apply(rename_Call_letter_status)

In [454]:
#Consider missing values as 'no'
df1["Call_letter_status"] = df1["Call_letter_status"].fillna(value='no')

In [455]:
df1["Call_letter_status"].value_counts(),df1["Call_letter_status"].value_counts().sum()

(yes    723
 no     217
 Name: Call_letter_status, dtype: int64,
 940)

In [456]:
df1["Call_letter_status"].isna().sum()

0

### Column 19."Ex_attend"

In [457]:
df1["Ex_attend"].value_counts(),df1["Ex_attend"].value_counts().sum()

(Yes          696
 Uncertain    183
 No            35
 NO            26
 Name: Ex_attend, dtype: int64,
 940)

In [458]:
df1=df1[~(df1["Ex_attend"]== '11:00 AM')]
df1=df1[~(df1["Ex_attend"]== '10.30 Am')]

In [459]:
df1["Ex_attend"].value_counts(),df1["Ex_attend"].value_counts().sum()

(Yes          696
 Uncertain    183
 No            35
 NO            26
 Name: Ex_attend, dtype: int64,
 940)

In [460]:
df1["Ex_attend"].isna().sum()

0

In [461]:
df1[df1["Ex_attend"]=="Uncertain"].shape

(183, 19)

In [462]:
#Merging "Ex_attend" values
def rename_Ex_attend(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value =='no' or value == 'uncertain':
        return value
    else:
        return 'yes'
    
df1.Ex_attend=df1.Ex_attend.apply(rename_Ex_attend)

In [463]:
df1["Ex_attend"].value_counts(),df1["Ex_attend"].value_counts().sum()

(yes          696
 uncertain    183
 no            61
 Name: Ex_attend, dtype: int64,
 940)

In [464]:
df1["Ex_attend"].isna().sum()

0

In [465]:
#Consider missing values as 'no'
df1["Ex_attend"]=df1["Ex_attend"].fillna(value="no")

In [466]:
df1["Ex_attend"].isna().sum()

0

In [467]:
df1["Ex_attend"].value_counts(),df1["Ex_attend"].value_counts().sum()

(yes          696
 uncertain    183
 no            61
 Name: Ex_attend, dtype: int64,
 940)

### Column 20."Obs_attend"

In [468]:
df1["Obs_attend"].value_counts(),df1["Obs_attend"].value_counts().sum()

(Yes     520
 No      314
 yes      66
 NO       26
 no        6
 No        6
 yes       1
 no        1
 Name: Obs_attend, dtype: int64,
 940)

In [469]:
df1["Obs_attend"].isna().sum()#No missing values

0

In [470]:
#Merging "Obs_attend" values
def rename_Obs_attend(data):
    if pd.isnull(data):
        return np.nan
    value=data.rstrip().lower()
    if value =='no':
        return value
    else:
        return 'yes'
    
df1.Obs_attend=df1.Obs_attend.apply(rename_Obs_attend)

In [471]:
df1["Obs_attend"].value_counts(),df1["Obs_attend"].value_counts().sum()

(yes    587
 no     353
 Name: Obs_attend, dtype: int64,
 940)

### Column 21."Matital_status"

In [472]:
df1["Matital_status"].value_counts(),df1["Matital_status"].value_counts().sum()

(Single     606
 Married    334
 Name: Matital_status, dtype: int64,
 940)

In [473]:
df1["Matital_status"].isna().sum() #NO misssing values

0

In [474]:
df1.isna().sum()/df1.shape[0]

Client_name          0.00000
Industry             0.00000
Closed_pos           0.00000
Interview_type       0.00000
Gender               0.00000
Curr_loc             0.00000
Curr_job_loc         0.00000
Interview_venue      0.00000
Permission_status    0.00000
Unscd_meet_status    0.00000
Followup_status      0.00000
AltNum_status        0.00000
Resume_status        0.00000
Venue_status         0.00000
Call_letter_status   0.00000
Ex_attend            0.00000
Obs_attend           0.00000
Matital_status       0.00000
Skill                0.00000
dtype: float64

In [475]:
for col in df1.columns:
    print("Unique values: ",col,"\n",df1[col].unique())

Unique values:  Client_name 
 ['Hospira' 'Aon Hewitt' 'UST' 'Standard Chartered Bank' 'ANZ'
 'Standard Chartered Bank Chennai' 'Aon hewitt Gurgaon' 'Astrazeneca'
 'Flextronics' 'Williams Lea' 'Hewitt']
Unique values:  Industry 
 ['Pharmaceuticals' 'IT' 'BFSI' 'Electronics']
Unique values:  Closed_pos 
 ['Non-Routine' 'Routine']
Unique values:  Interview_type 
 ['scheduled_walkin' 'scheduled' 'walkin']
Unique values:  Gender 
 ['Male' 'Female']
Unique values:  Curr_loc 
 ['Chennai' 'gurgaon' 'bangalore' 'delhi' 'noida']
Unique values:  Curr_job_loc 
 ['hosur' 'bangalore' 'chennai' 'gurgaon' 'noida']
Unique values:  Interview_venue 
 ['hosur' 'gurgaon' 'bangalore' 'chennai' 'noida']
Unique values:  Permission_status 
 ['yes' 'no']
Unique values:  Unscd_meet_status 
 ['yes' 'no']
Unique values:  Followup_status 
 ['yes' 'no']
Unique values:  AltNum_status 
 ['yes' 'no']
Unique values:  Resume_status 
 ['yes' 'no']
Unique values:  Venue_status 
 ['yes' 'no']
Unique values:  Call_letter_sta

In [476]:
for col in df1.columns:
    print(col," : ",df1[col].nunique())

Client_name  :  11
Industry  :  4
Closed_pos  :  2
Interview_type  :  3
Gender  :  2
Curr_loc  :  5
Curr_job_loc  :  5
Interview_venue  :  5
Permission_status  :  2
Unscd_meet_status  :  2
Followup_status  :  2
AltNum_status  :  2
Resume_status  :  2
Venue_status  :  2
Call_letter_status  :  2
Ex_attend  :  3
Obs_attend  :  2
Matital_status  :  2
Skill  :  7


# Saving the cleaned data as new .csv file

In [477]:
df1.to_csv('F:\Programming\Machine Learning\ClassWorks\project\interview(Cleaned).csv', index=False)

# Preparing Machine Learning Model

# Model 1: Logistic Regression

In [332]:
df=pd.read_csv('F:\Programming\Machine Learning\ClassWorks\project\interview(Cleaned).csv')

In [333]:
df[:5]

Client_name         Industry   Closed_pos    Interview_type Gender Curr_loc  \
0     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
1     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
2     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
3     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
4     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   

  Curr_job_loc Interview_venue Permission_status Unscd_meet_status  \
0        hosur           hosur               yes               yes   
1    bangalore           hosur               yes               yes   
2      chennai           hosur                no                no   
3      chennai           hosur               yes               yes   
4    bangalore           hosur               yes               yes   

  Followup_status AltNum_status Resume_status Venue_status Call_letter_status  \
0             yes           yes           yes          yes                yes   
1             yes           yes           yes          yes                yes   
2              no            no            no           no                 no   
3              no           yes            no          yes                yes   
4             yes           yes           yes          yes                yes   

   Ex_attend Obs_attend Matital_status    Skill  
0        yes         no         Single  Routine  
1        yes         no         Single  Routine  
2  uncertain         no         Single  Routine  
3  uncertain         no         Single  Routine  
4  uncertain         no        Married  Routine

In [334]:
df.shape

(940, 19)

In [335]:
cols=['Client_name', 'Industry', 'Closed_pos', 'Interview_type', 'Gender',
       'Curr_loc', 'Curr_job_loc', 'Interview_venue', 'Permission_status',
       'Unscd_meet_status', 'Followup_status', 'AltNum_status',
       'Resume_status', 'Venue_status', 'Call_letter_status', 'Ex_attend',
        'Matital_status', 'Skill']

In [336]:
cols

['Client_name',
 'Industry',
 'Closed_pos',
 'Interview_type',
 'Gender',
 'Curr_loc',
 'Curr_job_loc',
 'Interview_venue',
 'Permission_status',
 'Unscd_meet_status',
 'Followup_status',
 'AltNum_status',
 'Resume_status',
 'Venue_status',
 'Call_letter_status',
 'Ex_attend',
 'Matital_status',
 'Skill']

In [337]:
X=df.drop("Obs_attend",axis=1)
y=df["Obs_attend"]

## One hot encoding on categorical data

In [338]:
Xohe=pd.get_dummies(X,columns=cols)

In [339]:
df.shape

(940, 19)

In [340]:
Xohe.shape

(940, 63)

In [341]:
Xohe[:5]

Client_name_ANZ  Client_name_Aon Hewitt  Client_name_Aon hewitt Gurgaon  \
0                0                       0                               0   
1                0                       0                               0   
2                0                       0                               0   
3                0                       0                               0   
4                0                       0                               0   

   Client_name_Astrazeneca  Client_name_Flextronics  Client_name_Hewitt  \
0                        0                        0                   0   
1                        0                        0                   0   
2                        0                        0                   0   
3                        0                        0                   0   
4                        0                        0                   0   

   Client_name_Hospira  Client_name_Standard Chartered Bank  \
0                    1                                    0   
1                    1                                    0   
2                    1                                    0   
3                    1                                    0   
4                    1                                    0   

   Client_name_Standard Chartered Bank Chennai  Client_name_UST  ...  \
0                                            0                0  ...   
1                                            0                0  ...   
2                                            0                0  ...   
3                                            0                0  ...   
4                                            0                0  ...   

   Ex_attend_yes  Matital_status_Married  Matital_status_Single  \
0              1                       0                      1   
1              1                       0                      1   
2              0                       0                      1   
3              0                       0                      1   
4              0                       1                      0   

   Skill_Fresher  Skill_Java  Skill_KYC  Skill_Operations  Skill_Oracle  \
0              0           0          0                 0             0   
1              0           0          0                 0             0   
2              0           0          0                 0             0   
3              0           0          0                 0             0   
4              0           0          0                 0             0   

   Skill_Routine  Skill_Testing  
0              1              0  
1              1              0  
2              1              0  
3              1              0  
4              1              0  

[5 rows x 63 columns]

In [342]:
dict = {'yes' : 1, 'no' : 0}
y= y.map(dict)

In [343]:
## Target Class distribution:
y.value_counts()/y.shape[0]

1   0.62447
0   0.37553
Name: Obs_attend, dtype: float64

## Building Base Logistic Regression Model

In [344]:
Xtrain,Xtest,ytrain,ytest=model_selection.train_test_split(Xohe,y,test_size=.2,random_state=123)

In [345]:
model=linear_model.LogisticRegression()
model.fit(Xtrain,ytrain)

C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [346]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [347]:
print("-----Train data-----")
print("precision :",metrics.precision_score(ytrain,trainpred))
print("AUC :",metrics.roc_auc_score(ytrain,trainpred))
print("-----Test data-----")
print("precision :",metrics.precision_score(ytest,testpred))
print("AUC :",metrics.roc_auc_score(ytest,testpred))

-----Train data-----
precision : 0.7074010327022375
AUC : 0.6378129544101803
-----Test data-----
precision : 0.7412587412587412
AUC : 0.684866828087167


## P-Value check

In [173]:
import statsmodels.api as sm

In [174]:
Xtrain_1=sm.add_constant(Xtrain)

lrmodel=sm.GLM(ytrain,Xtrain_1,family=sm.families.Binomial() )

lrresult=lrmodel.fit()

C:\Users\user\anaconda3\lib\site-packages\statsmodels\genmod\families\links.py:188: RuntimeWarning: overflow encountered in exp
  t = np.exp(-z)


In [175]:
lrresult.summary()

C:\Users\user\anaconda3\lib\site-packages\statsmodels\genmod\families\family.py:947: RuntimeWarning: divide by zero encountered in log
  y * np.log(mu / (1 - mu + 1e-20)) +
C:\Users\user\anaconda3\lib\site-packages\statsmodels\genmod\families\family.py:947: RuntimeWarning: invalid value encountered in multiply
  y * np.log(mu / (1 - mu + 1e-20)) +


<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:             Obs_attend   No. Observations:                  752
Model:                            GLM   Df Residuals:                      716
Model Family:                Binomial   Df Model:                           35
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                    nan
Date:                Wed, 17 Nov 2021   Deviance:                       30486.
Time:                        19:34:15   Pearson chi2:                 1.49e+18
No. Iterations:                   100                                         
Covariance Type:            nonrobust                                         
===============================================================================================================
                                                  coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------------
const                                       -3.226e+14    2.7e+06   -1.2e+08      0.000   -3.23e+14   -3.23e+14
Client_name_ANZ                              8.117e+14   1.62e+07   5.01e+07      0.000    8.12e+14    8.12e+14
Client_name_Aon Hewitt                       -1.18e+15   2.66e+07  -4.44e+07      0.000   -1.18e+15   -1.18e+15
Client_name_Aon hewitt Gurgaon               7.161e+15   3.81e+07   1.88e+08      0.000    7.16e+15    7.16e+15
Client_name_Astrazeneca                      1.342e+14   1.62e+07   8.28e+06      0.000    1.34e+14    1.34e+14
Client_name_Flextronics                      2.301e+14   9.76e+06   2.36e+07      0.000     2.3e+14     2.3e+14
Client_name_Hewitt                          -4.391e+15   3.06e+07  -1.44e+08      0.000   -4.39e+15   -4.39e+15
Client_name_Hospira                         -7.365e+14   1.36e+07  -5.41e+07      0.000   -7.36e+14   -7.36e+14
Client_name_Standard Chartered Bank          2.143e+13   1.27e+07   1.69e+06      0.000    2.14e+13    2.14e+13
Client_name_Standard Chartered Bank Chennai -1.615e+15   1.68e+07  -9.59e+07      0.000   -1.61e+15   -1.61e+15
Client_name_UST                             -7.227e+14   3.09e+07  -2.34e+07      0.000   -7.23e+14   -7.23e+14
Client_name_Williams Lea                    -3.549e+13   4.68e+07  -7.59e+05      0.000   -3.55e+13   -3.55e+13
Industry_BFSI                               -7.817e+14   9.69e+06  -8.06e+07      0.000   -7.82e+14   -7.82e+14
Industry_Electronics                         2.301e+14   9.76e+06   2.36e+07      0.000     2.3e+14     2.3e+14
Industry_IT                                  8.313e+14   1.59e+07   5.24e+07      0.000    8.31e+14    8.31e+14
Industry_Pharmaceuticals                    -6.022e+14   1.05e+07  -5.75e+07      0.000   -6.02e+14   -6.02e+14
Closed_pos_Non-Routine                       1.429e+14   6.39e+06   2.24e+07      0.000    1.43e+14    1.43e+14
Closed_pos_Routine                          -4.655e+14   5.73e+06  -8.12e+07      0.000   -4.65e+14   -4.65e+14
Interview_type_scheduled                    -7.492e+13   8.96e+06  -8.36e+06      0.000   -7.49e+13   -7.49e+13
Interview_type_scheduled_walkin             -5.213e+14   5.99e+06  -8.71e+07      0.000   -5.21e+14   -5.21e+14
Interview_type_walkin                        2.736e+14   8.12e+06   3.37e+07      0.000    2.74e+14    2.74e+14
Gender_Female                               -2.206e+14   3.51e+06  -6.28e+07      0.000   -2.21e+14   -2.21e+14
Gender_Male                                  -1.02e+14   3.26e+06  -3.13e+07      0.000   -1.02e+14   -1.02e+14
Curr_loc_Chennai                            -1.496e+15    2.2e+07   -6.8e+07      0.000    -1.5e+15    -1.5e+15
Curr_loc_bangalore                           -4.16e+14   2.

## Performing RFE

In [176]:
model=linear_model.LogisticRegression()
model.fit(Xtrain,ytrain)

C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [177]:
#Recursive feature elimination
rfeobj=feature_selection.RFE(estimator=model,n_features_to_select=10)

In [178]:
rfeobj.fit(Xtrain,ytrain)

C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i 

RFE(estimator=LogisticRegression(), n_features_to_select=10)

In [179]:
Xtrain.columns[rfeobj.support_]

Index(['Client_name_ANZ', 'Client_name_Hospira', 'Interview_type_scheduled',
       'Interview_venue_hosur', 'Permission_status_no', 'Resume_status_no',
       'Venue_status_no', 'Ex_attend_no', 'Ex_attend_uncertain',
       'Ex_attend_yes'],
      dtype='object')

In [180]:
arr=rfeobj.transform(Xtrain)

In [181]:
Xtrain1=pd.DataFrame(arr,columns=Xtrain.columns[rfeobj.support_],index=Xtrain.index)

In [182]:
Xtrain1[:5]

Client_name_ANZ  Client_name_Hospira  Interview_type_scheduled  \
13                 0                    0                         0   
74                 0                    0                         0   
815                0                    0                         0   
895                0                    0                         0   
447                0                    0                         0   

     Interview_venue_hosur  Permission_status_no  Resume_status_no  \
13                       0                     1                 0   
74                       0                     0                 0   
815                      0                     0                 0   
895                      0                     0                 0   
447                      0                     0                 0   

     Venue_status_no  Ex_attend_no  Ex_attend_uncertain  Ex_attend_yes  
13                 0             0                    1              0  
74                 0             0                    0              1  
815                0             0                    0              1  
895                0             0                    0              1  
447                0             0                    0              1

In [183]:
Xtest1=pd.DataFrame(rfeobj.transform(Xtest),columns=Xtest.columns[rfeobj.support_],index=Xtest.index)

In [184]:
Xtest1[:5]

Client_name_ANZ  Client_name_Hospira  Interview_type_scheduled  \
770                0                    0                         1   
346                0                    0                         0   
558                0                    0                         0   
587                0                    0                         0   
274                0                    0                         0   

     Interview_venue_hosur  Permission_status_no  Resume_status_no  \
770                      0                     0                 0   
346                      0                     1                 1   
558                      0                     0                 0   
587                      0                     0                 0   
274                      0                     0                 0   

     Venue_status_no  Ex_attend_no  Ex_attend_uncertain  Ex_attend_yes  
770                0             0                    0              1  
346                1             0                    1              0  
558                0             0                    0              1  
587                0             0                    0              1  
274                0             0                    0              1

In [185]:
model=linear_model.LogisticRegression()
model.fit(Xtrain1,ytrain)

LogisticRegression()

In [186]:
trainpred=model.predict(Xtrain1)
testpred=model.predict(Xtest1)

## Metrics for the model

In [187]:
print("-----Train data-----")
print("precision :",metrics.precision_score(ytrain,trainpred))
print("AUC :",metrics.roc_auc_score(ytrain,trainpred))
print("-----Test data-----")
print("precision :",metrics.precision_score(ytest,testpred))
print("AUC :",metrics.roc_auc_score(ytest,testpred))

-----Train data-----
precision : 0.7102966841186736
AUC : 0.6406156998952738
-----Test data-----
precision : 0.7412587412587412
AUC : 0.684866828087167


In [188]:
## Final Columns to be chosen
Xtrain.columns[rfeobj.support_]

Index(['Client_name_ANZ', 'Client_name_Hospira', 'Interview_type_scheduled',
       'Interview_venue_hosur', 'Permission_status_no', 'Resume_status_no',
       'Venue_status_no', 'Ex_attend_no', 'Ex_attend_uncertain',
       'Ex_attend_yes'],
      dtype='object')

# Model2: Decision Tree

In [215]:
df=pd.read_csv("F:\Programming\Machine Learning\ClassWorks\project\interview(Cleaned).csv")

In [216]:
df.head(10)

Client_name         Industry   Closed_pos    Interview_type  Gender  \
0     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin    Male   
1     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin    Male   
2     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin    Male   
3     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin    Male   
4     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin    Male   
5  Aon Hewitt               IT  Non-Routine  scheduled_walkin    Male   
6  Aon Hewitt               IT  Non-Routine  scheduled_walkin    Male   
7  Aon Hewitt               IT  Non-Routine  scheduled_walkin  Female   
8  Aon Hewitt               IT  Non-Routine  scheduled_walkin    Male   
9  Aon Hewitt               IT  Non-Routine  scheduled_walkin  Female   

  Curr_loc Curr_job_loc Interview_venue Permission_status Unscd_meet_status  \
0  Chennai        hosur           hosur               yes               yes   
1  Chennai    bangalore           hosur               yes               yes   
2  Chennai      chennai           hosur                no                no   
3  Chennai      chennai           hosur               yes               yes   
4  Chennai    bangalore           hosur               yes               yes   
5  gurgaon      gurgaon         gurgaon               yes               yes   
6  gurgaon      gurgaon         gurgaon               yes               yes   
7  gurgaon      gurgaon         gurgaon               yes               yes   
8  gurgaon      gurgaon         gurgaon               yes               yes   
9  gurgaon      gurgaon         gurgaon               yes               yes   

  Followup_status AltNum_status Resume_status Venue_status Call_letter_status  \
0             yes           yes           yes          yes                yes   
1             yes           yes           yes          yes                yes   
2              no            no            no           no                 no   
3              no           yes            no          yes                yes   
4             yes           yes           yes          yes                yes   
5             yes           yes           yes          yes                yes   
6             yes           yes           yes          yes                yes   
7             yes           yes           yes          yes                yes   
8             yes           yes           yes          yes                yes   
9             yes           yes           yes          yes                yes   

   Ex_attend Obs_attend Matital_status    Skill  
0        yes         no         Single  Routine  
1        yes         no         Single  Routine  
2  uncertain         no         Single  Routine  
3  uncertain         no         Single  Routine  
4  uncertain         no        Married  Routine  
5        yes        yes         Single  Routine  
6        yes        yes         Single  Routine  
7        yes        yes         Single  Routine  
8        yes        yes         Single  Routine  
9        yes         no         Single  Routine

In [217]:
df.shape

(940, 19)

In [218]:
df["Ex_attend"]=df["Ex_attend"].replace({"yes":1,"no":0,"uncertain":2})

In [219]:
#Applying LabelEndoding on all the categorical columns
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
for column_name in df.columns:
    if df[column_name].dtype == object:
        df[column_name] = le.fit_transform(df[column_name])
    else:
        pass

In [220]:
#X->Data    
#y-> Target column : 1-> "will attend the interview" , 0-> "will NOT attend the interview"
X=df.drop("Obs_attend",axis=1)
y=df["Obs_attend"]

In [221]:
def printscores(act,pred):
    print("precision :",metrics.precision_score(act,pred))
    print("AUC :",metrics.roc_auc_score(act,pred))

In [222]:
df.head()

Client_name  Industry  Closed_pos  Interview_type  Gender  Curr_loc  \
0            6         3           0               1       1         0   
1            6         3           0               1       1         0   
2            6         3           0               1       1         0   
3            6         3           0               1       1         0   
4            6         3           0               1       1         0   

   Curr_job_loc  Interview_venue  Permission_status  Unscd_meet_status  \
0             3                3                  1                  1   
1             0                3                  1                  1   
2             1                3                  0                  0   
3             1                3                  1                  1   
4             0                3                  1                  1   

   Followup_status  AltNum_status  Resume_status  Venue_status  \
0                1              1              1             1   
1                1              1              1             1   
2                0              0              0             0   
3                0              1              0             1   
4                1              1              1             1   

   Call_letter_status  Ex_attend  Obs_attend  Matital_status  Skill  
0                   1          1           0               1      5  
1                   1          1           0               1      5  
2                   0          2           0               1      5  
3                   1          2           0               1      5  
4                   1          2           0               0      5

In [223]:
#Splitting into trainging and testing data
Xtrain,Xtest,ytrain,ytest=model_selection.train_test_split(X,y,test_size=.1,random_state=123)

## Applying Hyperparameter Tuning

##### hyperparameter tuning on 'max_depth'

In [224]:
grid={"max_depth":[1,2,3,4,5,6,7,8,9,10]}
model=tree.DecisionTreeClassifier(random_state=53)

In [225]:
gobj=model_selection.GridSearchCV(model,param_grid=grid,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [226]:
gobj.fit(Xtrain,ytrain)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=53), n_jobs=-1,
             param_grid={'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},
             return_train_score=True, scoring='precision')

In [227]:
gobj.best_params_

{'max_depth': 2}

In [228]:
gobjdf=pd.DataFrame(gobj.cv_results_)

In [229]:
gobjdf.shape

(10, 21)

In [230]:
gobjdf[["params","mean_train_score","mean_test_score"]]

params  mean_train_score  mean_test_score
0   {'max_depth': 1}           0.66203          0.66209
1   {'max_depth': 2}           0.71018          0.70672
2   {'max_depth': 3}           0.70521          0.68522
3   {'max_depth': 4}           0.71657          0.69373
4   {'max_depth': 5}           0.71484          0.69179
5   {'max_depth': 6}           0.71362          0.68607
6   {'max_depth': 7}           0.72429          0.68415
7   {'max_depth': 8}           0.73196          0.68168
8   {'max_depth': 9}           0.73687          0.67770
9  {'max_depth': 10}           0.73918          0.67853

In [231]:
model=tree.DecisionTreeClassifier(max_depth=2,random_state=53)
model.fit(Xtrain,ytrain)

DecisionTreeClassifier(max_depth=2, random_state=53)

In [232]:
#Prediction on training and testing data
#Prediction on training and testing data
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)
print("Training data: ")
printscores(ytrain,trainpred)
print("\nTest data: ")
printscores(ytest,testpred)

Training data: 
precision : 0.7110754414125201
AUC : 0.64320681807406

Test data: 
precision : 0.7671232876712328
AUC : 0.6702508960573477


##### hyperparameter tuning on "min_samples_split"  and  "min_samples_leaf"

In [233]:
grid={"min_samples_split":[2,3,4,5],"min_samples_leaf":[18,19,20,21,22,23]}
model=tree.DecisionTreeClassifier(random_state=53)

In [234]:
gobj=model_selection.GridSearchCV(model,param_grid=grid,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [235]:
gobj.fit(Xtrain,ytrain)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=53), n_jobs=-1,
             param_grid={'min_samples_leaf': [18, 19, 20, 21, 22, 23],
                         'min_samples_split': [2, 3, 4, 5]},
             return_train_score=True, scoring='precision')

In [236]:
gobj.best_params_

{'min_samples_leaf': 19, 'min_samples_split': 2}

In [237]:
gobjdf=pd.DataFrame(gobj.cv_results_)

In [238]:
gobjdf.shape

(24, 22)

In [239]:
gobjdf[["params","mean_train_score","mean_test_score"]]

params  mean_train_score  \
0   {'min_samples_leaf': 18, 'min_samples_split': 2}           0.70454   
1   {'min_samples_leaf': 18, 'min_samples_split': 3}           0.70454   
2   {'min_samples_leaf': 18, 'min_samples_split': 4}           0.70454   
3   {'min_samples_leaf': 18, 'min_samples_split': 5}           0.70454   
4   {'min_samples_leaf': 19, 'min_samples_split': 2}           0.70606   
5   {'min_samples_leaf': 19, 'min_samples_split': 3}           0.70606   
6   {'min_samples_leaf': 19, 'min_samples_split': 4}           0.70606   
7   {'min_samples_leaf': 19, 'min_samples_split': 5}           0.70606   
8   {'min_samples_leaf': 20, 'min_samples_split': 2}           0.70485   
9   {'min_samples_leaf': 20, 'min_samples_split': 3}           0.70485   
10  {'min_samples_leaf': 20, 'min_samples_split': 4}           0.70485   
11  {'min_samples_leaf': 20, 'min_samples_split': 5}           0.70485   
12  {'min_samples_leaf': 21, 'min_samples_split': 2}           0.70145   
13  {'min_samples_leaf': 21, 'min_samples_split': 3}           0.70145   
14  {'min_samples_leaf': 21, 'min_samples_split': 4}           0.70145   
15  {'min_samples_leaf': 21, 'min_samples_split': 5}           0.70145   
16  {'min_samples_leaf': 22, 'min_samples_split': 2}           0.70094   
17  {'min_samples_leaf': 22, 'min_samples_split': 3}           0.70094   
18  {'min_samples_leaf': 22, 'min_samples_split': 4}           0.70094   
19  {'min_samples_leaf': 22, 'min_samples_split': 5}           0.70094   
20  {'min_samples_leaf': 23, 'min_samples_split': 2}           0.70204   
21  {'min_samples_leaf': 23, 'min_samples_split': 3}           0.70204   
22  {'min_samples_leaf': 23, 'min_samples_split': 4}           0.70204   
23  {'min_samples_leaf': 23, 'min_samples_split': 5}           0.70204   

    mean_test_score  
0           0.68620  
1           0.68620  
2           0.68620  
3           0.68620  
4           0.69106  
5           0.69106  
6           0.69106  
7           0.69106  
8           0.69060  
9           0.69060  
10          0.69060  
11          0.69060  
12          0.68161  
13          0.68161  
14          0.68161  
15          0.68161  
16          0.68334  
17          0.68334  
18          0.68334  
19          0.68334  
20          0.68171  
21          0.68171  
22          0.68171  
23          0.68171

In [240]:
model=tree.DecisionTreeClassifier(min_samples_leaf=19,min_samples_split=2,random_state=53)
model.fit(Xtrain,ytrain)

DecisionTreeClassifier(min_samples_leaf=19, random_state=53)

In [241]:
#Prediction on training and testing data
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)
print("Training data: ")
printscores(ytrain,trainpred)
print("\nTest data: ")
printscores(ytest,testpred)

Training data: 
precision : 0.7101226993865031
AUC : 0.6483156322602057

Test data: 
precision : 0.76
AUC : 0.6620583717357911


#### Now we are going to use RFE through GridSearch

In [242]:
model=tree.DecisionTreeClassifier(max_depth=2,min_samples_leaf=19,min_samples_split=2,random_state=53)

In [243]:
from sklearn import feature_selection
rfeobj=feature_selection.RFE(estimator=model)

In [244]:
grid={ 'n_features_to_select': [1,2,3,4,5,6,7,8,9]}
gobj=model_selection.GridSearchCV(return_train_score=True,estimator=rfeobj,param_grid=grid,cv=5,scoring="precision",n_jobs=-1)

In [245]:
gobj.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RFE(estimator=DecisionTreeClassifier(max_depth=2,
                                                            min_samples_leaf=19,
                                                            random_state=53)),
             n_jobs=-1,
             param_grid={'n_features_to_select': [1, 2, 3, 4, 5, 6, 7, 8, 9]},
             return_train_score=True, scoring='precision')

In [246]:
gobj.best_params_

{'n_features_to_select': 1}

In [247]:
gdf=pd.DataFrame(gobj.cv_results_)

In [248]:
gdf[ ["params","mean_train_score","mean_test_score"] ]

params  mean_train_score  mean_test_score
0  {'n_features_to_select': 1}           0.71110          0.71142
1  {'n_features_to_select': 2}           0.71018          0.70672
2  {'n_features_to_select': 3}           0.71018          0.70672
3  {'n_features_to_select': 4}           0.71018          0.70672
4  {'n_features_to_select': 5}           0.71018          0.70672
5  {'n_features_to_select': 6}           0.71018          0.70672
6  {'n_features_to_select': 7}           0.71018          0.70672
7  {'n_features_to_select': 8}           0.71018          0.70672
8  {'n_features_to_select': 9}           0.71018          0.70672

In [249]:
#Though we are getting n_features_to_select=1 , it is better to consider more than 1 column.
#Since Train and Test score remain same for n_features_to_select>1.
#Lets consider n_features_to_select=6
rfeobj=feature_selection.RFE(estimator=model,n_features_to_select=6)

In [250]:
rfeobj.fit(Xtrain,ytrain)

RFE(estimator=DecisionTreeClassifier(max_depth=2, min_samples_leaf=19,
                                     random_state=53),
    n_features_to_select=6)

In [251]:
Xtrain.columns[rfeobj.support_]

Index(['Gender', 'Venue_status', 'Call_letter_status', 'Ex_attend',
       'Matital_status', 'Skill'],
      dtype='object')

In [252]:
arr1=rfeobj.transform(Xtrain)
arr2=rfeobj.transform(Xtest)

In [253]:
Xtrain3=pd.DataFrame(arr1,columns=Xtrain.columns[rfeobj.support_],index=Xtrain.index)
Xtest3=pd.DataFrame(arr2,columns=Xtest.columns[rfeobj.support_],index=Xtest.index)

In [254]:
model=tree.DecisionTreeClassifier(max_depth=2,min_samples_leaf=19,min_samples_split=2,random_state=53)
model.fit(Xtrain3,ytrain)

DecisionTreeClassifier(max_depth=2, min_samples_leaf=19, random_state=53)

In [255]:
#Prediction on training and testing data
trainpred=model.predict(Xtrain3)
testpred=model.predict(Xtest3)
print("Training data: ")
printscores(ytrain,trainpred)
print("\nTest data: ")
printscores(ytest,testpred)

Training data: 
precision : 0.7110754414125201
AUC : 0.64320681807406

Test data: 
precision : 0.7671232876712328
AUC : 0.6702508960573477


In [258]:
## Final Columns to be chosen
Xtrain.columns[rfeobj.support_]

Index(['Gender', 'Venue_status', 'Call_letter_status', 'Ex_attend',
       'Matital_status', 'Skill'],
      dtype='object')

# Model 3: Random Forest

In [259]:
df1=pd.read_csv('F:\Programming\Machine Learning\ClassWorks\project\interview(Cleaned).csv')

In [260]:
df1[:5]

Client_name         Industry   Closed_pos    Interview_type Gender Curr_loc  \
0     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
1     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
2     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
3     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
4     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   

  Curr_job_loc Interview_venue Permission_status Unscd_meet_status  \
0        hosur           hosur               yes               yes   
1    bangalore           hosur               yes               yes   
2      chennai           hosur                no                no   
3      chennai           hosur               yes               yes   
4    bangalore           hosur               yes               yes   

  Followup_status AltNum_status Resume_status Venue_status Call_letter_status  \
0             yes           yes           yes          yes                yes   
1             yes           yes           yes          yes                yes   
2              no            no            no           no                 no   
3              no           yes            no          yes                yes   
4             yes           yes           yes          yes                yes   

   Ex_attend Obs_attend Matital_status    Skill  
0        yes         no         Single  Routine  
1        yes         no         Single  Routine  
2  uncertain         no         Single  Routine  
3  uncertain         no         Single  Routine  
4  uncertain         no        Married  Routine

# Converting the categorical values into numerical

In [261]:
dict1 = {'Standard Chartered Bank' : 1, 'Hospira' : 2, 'ANZ' : 3, 'Flextronics' : 4, 'UST' : 5, 'Standard Chartered Bank Chennai' : 6, 'Aon Hewitt' : 7, 'Astrazeneca' : 8, 'Hewitt' : 9, 'Williams Lea' : 10, 'Aon hewitt Gurgaon' : 11}
dict2 = {'BFSI' : 1, 'Pharmaceuticals' : 2, 'IT' : 3, 'Electronics' : 4}
dict3 = {'Routine' : 1, 'Non-Routine' : 2}
dict4 = {'scheduled_walkin' : 1, 'walkin' : 2, 'scheduled' : 3}
dict5 = {'Male' : 1, 'Female' : 2}
dict6 = {'Chennai' : 1, 'bangalore' : 2, 'gurgaon' : 3, 'noida' : 4, 'delhi' : 5}
dict7 = {'chennai' : 1, 'bangalore' : 2, 'gurgaon' : 3, 'noida' : 4, 'hosur' : 5}
dict8 = {'yes' : 1, 'no' : 0}
dict9 = {'yes' : 1, 'no' : 2, "uncertain":3}
dict10 = {'Single' : 1, 'Married' : 2}
dict11 = {'Java' : 1, 'KYC' : 2, 'Operations' : 3, 'Fresher' : 4, 'Oracle' : 5, 'Routine' : 6, 'Testing' : 7}

df1['Client_name']= df1['Client_name'].map(dict1)
df1['Industry']= df1['Industry'].map(dict2)
df1['Closed_pos']= df1['Closed_pos'].map(dict3)
df1['Interview_type']= df1['Interview_type'].map(dict4)
df1['Gender']= df1['Gender'].map(dict5)
df1['Curr_loc']= df1['Curr_loc'].map(dict6)
df1['Curr_job_loc']= df1['Curr_job_loc'].map(dict7)
df1['Interview_venue']= df1['Interview_venue'].map(dict7)
df1['Permission_status']= df1['Permission_status'].map(dict8)
df1['Unscd_meet_status']= df1['Unscd_meet_status'].map(dict8)
df1['Followup_status']= df1['Followup_status'].map(dict8)
df1['AltNum_status']= df1['AltNum_status'].map(dict8)
df1['Resume_status']= df1['Resume_status'].map(dict8)
df1['Venue_status']= df1['Venue_status'].map(dict8)
df1['Call_letter_status']= df1['Call_letter_status'].map(dict8)
df1['Ex_attend']= df1['Ex_attend'].map(dict9)
df1['Matital_status']= df1['Matital_status'].map(dict10)
df1['Obs_attend']= df1['Obs_attend'].map(dict8)
df1['Skill']= df1['Skill'].map(dict11)



In [262]:
df1[:5]

Client_name  Industry  Closed_pos  Interview_type  Gender  Curr_loc  \
0            2         2           2               1       1         1   
1            2         2           2               1       1         1   
2            2         2           2               1       1         1   
3            2         2           2               1       1         1   
4            2         2           2               1       1         1   

   Curr_job_loc  Interview_venue  Permission_status  Unscd_meet_status  \
0             5                5                  1                  1   
1             2                5                  1                  1   
2             1                5                  0                  0   
3             1                5                  1                  1   
4             2                5                  1                  1   

   Followup_status  AltNum_status  Resume_status  Venue_status  \
0                1              1              1             1   
1                1              1              1             1   
2                0              0              0             0   
3                0              1              0             1   
4                1              1              1             1   

   Call_letter_status  Ex_attend  Obs_attend  Matital_status  Skill  
0                   1          1           0               1      6  
1                   1          1           0               1      6  
2                   0          3           0               1      6  
3                   1          3           0               1      6  
4                   1          3           0               2      6

# Building Random Forest Model

In [263]:
X=df1.drop("Obs_attend",axis=1)
y=df1["Obs_attend"]
Xtrain,Xtest,ytrain,ytest=model_selection.train_test_split(X,y,test_size=.1,random_state=123)

In [264]:
model=ensemble.RandomForestClassifier(random_state=42) 

In [265]:
trees=list()
for i in range(1,150):
    trees.append(i)

## Hyperparameter tuning

In [266]:
grid1={ "max_depth":[1,2,3,4,5,6,7,8,9,10,11,12]}
grid2={ "max_features": ['auto', 'sqrt', 'log2']}
grid3={ "n_estimators" : trees}
grid4={"min_samples_split":[2,3,4,5]}
grid5={"min_samples_leaf":[10,11,12,13,14]}

### For max_depth

In [267]:
gobj=model_selection.GridSearchCV(model,param_grid=grid1,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [268]:
gobj.fit(Xtrain,ytrain)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]},
             return_train_score=True, scoring='precision')

In [269]:
gobj.best_params_

{'max_depth': 3}

In [270]:
#model=ensemble.RandomForestClassifier(n_estimators=60,max_features=8,max_depth=2,random_state=42) 
model=ensemble.RandomForestClassifier(max_depth=3,random_state=42) 

In [271]:
model.fit(Xtrain,ytrain)

RandomForestClassifier(max_depth=3, random_state=42)

In [272]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [273]:
print("train precision :",metrics.precision_score(ytrain,trainpred))
print("test precision :",metrics.precision_score(ytest,testpred))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred))

train precision : 0.7115689381933439
test precision : 0.7702702702702703

AUC for train : 0.6458264188516429
AUC for test : 0.6781874039938556


### For n_estimators

In [274]:
gobj_estimators=model_selection.GridSearchCV(model,param_grid=grid3,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [275]:
gobj_estimators.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(max_depth=3, random_state=42),
             n_jobs=-1,
             param_grid={'n_estimators': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
                                          13, 14, 15, 16, 17, 18, 19, 20, 21,
                                          22, 23, 24, 25, 26, 27, 28, 29, 30, ...]},
             return_train_score=True, scoring='precision')

In [276]:
gobj_estimators.best_params_

{'n_estimators': 57}

In [277]:
model=ensemble.RandomForestClassifier(n_estimators=57,max_depth=3,random_state=42) 

In [278]:
model.fit(Xtrain,ytrain)

RandomForestClassifier(max_depth=3, n_estimators=57, random_state=42)

In [279]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [280]:
print("train precision :",metrics.precision_score(ytrain,trainpred))
print("test precision :",metrics.precision_score(ytest,testpred))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred))

train precision : 0.7133757961783439
test precision : 0.7702702702702703

AUC for train : 0.6479778104404723
AUC for test : 0.6781874039938556


### For max_features

In [281]:
gobj_max_features=model_selection.GridSearchCV(model,param_grid=grid2,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [282]:
gobj_max_features.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(max_depth=3, n_estimators=57,
                                              random_state=42),
             n_jobs=-1, param_grid={'max_features': ['auto', 'sqrt', 'log2']},
             return_train_score=True, scoring='precision')

In [283]:
gobj_max_features.best_params_

{'max_features': 'auto'}

In [284]:
model=ensemble.RandomForestClassifier(max_depth=3,n_estimators=57,max_features='auto',random_state=42) 

In [285]:
model.fit(Xtrain,ytrain)

RandomForestClassifier(max_depth=3, n_estimators=57, random_state=42)

In [286]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [287]:
print("train precision :",metrics.precision_score(ytrain,trainpred))
print("test precision :",metrics.precision_score(ytest,testpred))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred))

train precision : 0.7133757961783439
test precision : 0.7702702702702703

AUC for train : 0.6479778104404723
AUC for test : 0.6781874039938556


### For min_samples_split

In [288]:
gobj_min_samples_split=model_selection.GridSearchCV(model,param_grid=grid4,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [289]:
gobj_min_samples_split.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(max_depth=3, n_estimators=57,
                                              random_state=42),
             n_jobs=-1, param_grid={'min_samples_split': [2, 3, 4, 5]},
             return_train_score=True, scoring='precision')

In [290]:
gobj_min_samples_split.best_params_

{'min_samples_split': 2}

In [291]:
model=ensemble.RandomForestClassifier(max_depth=3,n_estimators=57,max_features='auto',min_samples_split=2,random_state=42) 

In [292]:
model.fit(Xtrain,ytrain)

RandomForestClassifier(max_depth=3, n_estimators=57, random_state=42)

In [293]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [294]:
print("train precision :",metrics.precision_score(ytrain,trainpred))
print("test precision :",metrics.precision_score(ytest,testpred))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred))

train precision : 0.7133757961783439
test precision : 0.7702702702702703

AUC for train : 0.6479778104404723
AUC for test : 0.6781874039938556


### For min_samples_leaf

In [295]:
gobj_min_samples_leaf=model_selection.GridSearchCV(model,param_grid=grid5,cv=5,scoring="precision",n_jobs=-1,return_train_score=True)

In [296]:
gobj_min_samples_leaf.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RandomForestClassifier(max_depth=3, n_estimators=57,
                                              random_state=42),
             n_jobs=-1, param_grid={'min_samples_leaf': [10, 11, 12, 13, 14]},
             return_train_score=True, scoring='precision')

In [297]:
gobj_min_samples_leaf.best_params_

{'min_samples_leaf': 10}

In [298]:
model=ensemble.RandomForestClassifier(max_depth=3,n_estimators=57,max_features='auto',min_samples_leaf=10,min_samples_split=2,random_state=42) 

In [299]:
model.fit(Xtrain,ytrain)

RandomForestClassifier(max_depth=3, min_samples_leaf=10, n_estimators=57,
                       random_state=42)

In [300]:
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)

In [301]:
print("train precision :",metrics.precision_score(ytrain,trainpred))
print("test precision :",metrics.precision_score(ytest,testpred))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred))

train precision : 0.7106518282988871
test precision : 0.7702702702702703

AUC for train : 0.643918021905078
AUC for test : 0.6781874039938556


## Feature extraction using RFE

In [302]:
model=ensemble.RandomForestClassifier(max_depth=3,n_estimators=57,max_features='auto',min_samples_leaf=10,min_samples_split=2,random_state=42) 

In [303]:
rfeobj=feature_selection.RFE(estimator=model)

In [304]:
grid={ 'n_features_to_select': [3,4,5,6,7,8,9]}
gobj2=model_selection.GridSearchCV(return_train_score=True,estimator=rfeobj,param_grid=grid,cv=5,scoring="precision",n_jobs=-1)

In [305]:
gobj2.fit(Xtrain,ytrain)

GridSearchCV(cv=5,
             estimator=RFE(estimator=RandomForestClassifier(max_depth=3,
                                                            min_samples_leaf=10,
                                                            n_estimators=57,
                                                            random_state=42)),
             n_jobs=-1,
             param_grid={'n_features_to_select': [3, 4, 5, 6, 7, 8, 9]},
             return_train_score=True, scoring='precision')

In [306]:
gobj2.best_params_

{'n_features_to_select': 5}

In [307]:
rfeobj=feature_selection.RFE(estimator=model,n_features_to_select=5)

In [308]:
rfeobj.fit(Xtrain,ytrain)

RFE(estimator=RandomForestClassifier(max_depth=3, min_samples_leaf=10,
                                     n_estimators=57, random_state=42),
    n_features_to_select=5)

In [309]:
Xtrain.columns[rfeobj.support_]

Index(['Permission_status', 'Resume_status', 'Venue_status',
       'Call_letter_status', 'Ex_attend'],
      dtype='object')

In [310]:
Xtrain1=Xtrain[Xtrain.columns[rfeobj.support_]]
Xtest1=Xtest[Xtest.columns[rfeobj.support_]]

In [311]:
Xtrain1[:5]

Permission_status  Resume_status  Venue_status  Call_letter_status  \
50                   1              1             1                   1   
807                  0              0             0                   0   
147                  1              1             1                   1   
602                  1              1             1                   1   
866                  1              1             1                   1   

     Ex_attend  
50           1  
807          3  
147          1  
602          1  
866          1

In [312]:
Xtest1[:5]

Permission_status  Resume_status  Venue_status  Call_letter_status  \
770                  1              1             1                   1   
346                  0              0             0                   0   
558                  1              1             1                   1   
587                  1              1             1                   1   
274                  1              1             1                   1   

     Ex_attend  
770          1  
346          3  
558          1  
587          1  
274          1

In [313]:
model=ensemble.RandomForestClassifier(max_depth=3,n_estimators=57,max_features='auto',min_samples_leaf=10,min_samples_split=2,random_state=42)  
model.fit(Xtrain1,ytrain)

RandomForestClassifier(max_depth=3, min_samples_leaf=10, n_estimators=57,
                       random_state=42)

In [314]:
trainpred_final=model.predict(Xtrain1)
testpred_final=model.predict(Xtest1)

#### Metrics obtained for the model final Random forest model

In [315]:
print("train precision :",metrics.precision_score(ytrain,trainpred_final))
print("test precision :",metrics.precision_score(ytest,testpred_final))

print("\nAUC for train :",metrics.roc_auc_score(ytrain,trainpred_final))
print("AUC for test :",metrics.roc_auc_score(ytest,testpred_final))

train precision : 0.7110754414125201
test precision : 0.7671232876712328

AUC for train : 0.64320681807406
AUC for test : 0.6702508960573477


# Model 4: Naive bayes

In [316]:
from sklearn import naive_bayes

In [317]:
df=pd.read_csv('F:\Programming\Machine Learning\ClassWorks\project\interview(Cleaned).csv')

In [318]:
df.head()

Client_name         Industry   Closed_pos    Interview_type Gender Curr_loc  \
0     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
1     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
2     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
3     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   
4     Hospira  Pharmaceuticals  Non-Routine  scheduled_walkin   Male  Chennai   

  Curr_job_loc Interview_venue Permission_status Unscd_meet_status  \
0        hosur           hosur               yes               yes   
1    bangalore           hosur               yes               yes   
2      chennai           hosur                no                no   
3      chennai           hosur               yes               yes   
4    bangalore           hosur               yes               yes   

  Followup_status AltNum_status Resume_status Venue_status Call_letter_status  \
0             yes           yes           yes          yes                yes   
1             yes           yes           yes          yes                yes   
2              no            no            no           no                 no   
3              no           yes            no          yes                yes   
4             yes           yes           yes          yes                yes   

   Ex_attend Obs_attend Matital_status    Skill  
0        yes         no         Single  Routine  
1        yes         no         Single  Routine  
2  uncertain         no         Single  Routine  
3  uncertain         no         Single  Routine  
4  uncertain         no        Married  Routine

In [319]:
model=naive_bayes.GaussianNB()

In [320]:
#Splitting into trainging and testing data
Xtrain,Xtest,ytrain,ytest=model_selection.train_test_split(X,y,test_size=.1,random_state=123)

In [321]:
model.fit(Xtrain,ytrain)

GaussianNB()

In [322]:
def printscores(act,pred):
    print("precision :",metrics.precision_score(act,pred))
    print("AUC :",metrics.roc_auc_score(act,pred))

In [323]:
#Prediction on training and testing data
trainpred=model.predict(Xtrain)
testpred=model.predict(Xtest)
print("Training data: ")
printscores(ytrain,trainpred)
print("\nTest data: ")
printscores(ytest,testpred)

Training data: 
precision : 0.6987577639751553
AUC : 0.6281470769522546

Test data: 
precision : 0.75
AUC : 0.6459293394777266


# Final Summary

### All the results for each models are as follows

### 1.Logistic regression 

Training data:
    
    precision : 0.7102966841186736
    AUC : 0.6406156998952738

Test data: 

    precision : 0.7412587412587412
    AUC : 0.684866828087167

Final Columns to be chosen:

{'Client_name_ANZ', 'Client_name_Hospira', 'Interview_type_scheduled',
'Interview_venue_hosur', 'Permission_status_no', 'Resume_status_no',
'Venue_status_no', 'Ex_attend_no', 'Ex_attend_uncertain',
'Ex_attend_yes'}

## 2.Decision Tree 

Training data: 

    precision : 0.7110754414125201

    AUC : 0.64320681807406

Test data: 

    precision : 0.7671232876712328

    AUC : 0.6702508960573477
    
Final Columns to be chosen:
    
{'Gender', 'Venue_status', 'Call_letter_status', 'Ex_attend',
'Matital_status', 'Skill'}

## 3.Random Forest

Training data:
    
    precision : 0.7110754414125201
    AUC: 0.64320681807406

Test data:
    
    precision : 0.7671232876712328
    AUC: 0.6702508960573477
    
Final Columns to be chosen
{'Permission_status', 'Resume_status', 'Venue_status',
'Call_letter_status', 'Ex_attend'}

## 4.Naive Bayes

Training data: 

    precision : 0.6987577639751553
    AUC : 0.6281470769522546

Test data: 

    precision : 0.75
    AUC : 0.6459293394777266

# Conclusion

### For the given data Decision Tree (or) Random Forest are the best models